docs/pallas/gpu/pipelining.md

```diff
@@ -0,0 +1,325 @@
+- `dimension_semantics`: A tuple of `Literal['parallel', 'sequential']` that specifies iteration semantics for each grid dimension. `parallel` will partition the corresponding dimension over the CUDA grid, and `sequential` dimensions will be pipelined sequentially. **Note that if no dimensions are marked `sequential`, no pipelining will happen!**
+- `max_concurrent_steps`: the number of pipeline stages; identical to the option in `emit_pipeline`.
+- `delay_release`: the number of steps to wait before buffer re-use; idential to the option in `emit_pipeline`.
+
+We recommend that users use `plgpu.kernel` rather than `pl.pallas_call` as `plgpu.kernel` supports more features (such as specifying the number of warpgroups and warp specialization).
+
+<!-- #endregion -->
+
+<!-- #region id="Qp3X6wylJtoa" -->
+### GPU Memory Spaces
+
+Refs can exist in one of 3 memory spaces, specified by passing in the appropriate memory space into the `BlockSpec`, i.e. `BlockSpec(memory_space=plgpu.GPUMemorySpace.GMEM)`.
+
+- `plgpu.GPUMemorySpace.SMEM` allocates a Ref in Shared Memory (SMEM). SMEM Refs can be dereferenced using array indexing syntax to store values in registers for compute, i.e. `x = y_ref[...]`.
+
+- `plgpu.GPUMemorySpace.GMEM` allocates a Ref in Global Memory (GMEM/HBM). Any Refs allocated in GMEM are not pipelined, and values cannot be accessed directly using array indexing operations. Instead, GMEM must be accessed via SMEM using `plgpu.copy_gmem_to_smem` for reading, or `plgpu.copy_smem_to_gmem` for writing, or pipelined into SMEM using `plgpu.emit_pipeline`.
+
+- `plgpu.GPUMemorySpace.TMEM` allocates a Ref in Tensor Memory (TMEM). This is a TensorCore Gen 5 (Blackwell) feature for storing arguments and results of asynchronous matrix multiplication instructions. Like GMEM, TMEM cannot be accessed directly but must first be copied to/from SMEM.
+
+The primary purpose of `emit_pipeline` is used to overlap TensorCore computation with data transfers between GMEM and SMEM, since asynchronous copies between GMEM/SMEM have a long latency, but all TensorCore computation must operate on registers (or TMEM/SMEM Refs in the case of matrix multiplication).
+<!-- #endregion -->
+
+<!-- #region id="0uzcrDCtKABQ" -->
+### Example: Matmul Kernel on Hopper GPUs
+<!-- #endregion -->
+
+<!-- #region id="vILVdlqEdoEK" -->
+Let's begin with a matrix multiplication example designed to run on Hopper GPUs. This kernel utilizes the Hopper-specific `wgmma` (warpgroup matrix multiply accumulate) instruction, which issues a matrix multiplication that runs asynchronously on the TensorCore. Under the hood, `wgmma` is a warpgroup-level instruction in which all 128 CUDA threads in a warpgroup collectively issue a matrix multiplication command. But since a program "thread" in Mosaic GPU conveniently corresponds to a warpgroup, all details of `wgmma` at the CUDA thread level are abstracted away.
```

Comment:
This is way too low-level and all those details are explained in the reference already. Just say that `wgmma` is issued by a single Pallas thread.

Reply:
Done.